### ISSUE TITLE
grid-row-span and grid-column-span not being prefixed by autoprefixer

### ISSUE THREAD

**Issue by franktopel** [postcss/autoprefixer#1063]
I have this SCSS mixin which creates css rules:

```
$columns: 12;
$rows: 12;

@mixin colspan-x {
    @for $i from 1 through $columns {
      .colspan#{$i} { grid-column-span: $i; }
    }
  }

  @mixin rowspan-x {
    @for $i from 1 through $rows {
      .rowspan#{$i} { grid-row-span: $i; }
    }
  }

@include colspan-x;
@include rowspan-x;
```
This creates

```
.colspan1 {
  grid-column-span: 1;
}

.colspan2 {
  grid-column-span: 2;
}

.colspan3 {
  grid-column-span: 3;
}

.colspan4 {
  grid-column-span: 4;
}

.colspan5 {
  grid-column-span: 5;
}

.colspan6 {
  grid-column-span: 6;
}

.colspan7 {
  grid-column-span: 7;
}

.colspan8 {
  grid-column-span: 8;
}

.colspan9 {
  grid-column-span: 9;
}

.colspan10 {
  grid-column-span: 10;
}

.colspan11 {
  grid-column-span: 11;
}

.colspan12 {
  grid-column-span: 12;
}

.rowspan1 {
  grid-row-span: 1;
}

.rowspan2 {
  grid-row-span: 2;
}

.rowspan3 {
  grid-row-span: 3;
}

.rowspan4 {
  grid-row-span: 4;
}

.rowspan5 {
  grid-row-span: 5;
}

.rowspan6 {
  grid-row-span: 6;
}

.rowspan7 {
  grid-row-span: 7;
}

.rowspan8 {
  grid-row-span: 8;
}

.rowspan9 {
  grid-row-span: 9;
}

.rowspan10 {
  grid-row-span: 10;
}

.rowspan11 {
  grid-row-span: 11;
}

.rowspan12 {
  grid-row-span: 12;
}
```

If I put this through autoprefixer (grid: true) the output stays identical, no `-ms-grid-column-span` and `-ms-grid-row-span` properties are being added. Other properties like `grid-column` and `grid-row` are properly prefixed.

**Comment by ai** [comment]
Check Autoprefixer warning. Grid support is limited and requires some additional properties.

How do you run Autopeprefixer?

**Comment by franktopel** [comment]
Autoprefixer is run as a part of the new @vue/cli build process.

**Comment by ai** [comment]
Try this code from [tests](https://github.com/postcss/autoprefixer/blob/28a4587cfa3b101e31b0e402141a507299799ba7/test/autoprefixer.test.js#L496):

```css
.warn {
    grid-column-end: 3;
    grid: subgrid;
    grid-template-areas: "head head"
                         "nav  main"
                         "foot ....";
}
```

**Comment by franktopel** [comment]
No warnings on the terminal I run the dev server in.

This is being created in my vue demo project:

```
.warn {
  -ms-grid-column-span: 3;
  grid-column-end: 3;
  grid: subgrid;
  grid-template-areas: "head head" "nav  main" "foot ....";
}
```

This is created by https://autoprefixer.github.io/ using `last 1 version, ie <= 11` as a filter:

```
.warn {
    grid-column-end: 3;
    grid: subgrid;
        grid-template-areas: "head head"
                         "nav  main"
                         "foot ....";
}
```

**Comment by ai** [comment]
Can I ask you to create an issue in related Vue project (sorry, I have no idea where to create it). Mention me there, I will come and describe details.

Unfortunately, https://autoprefixer.github.io/ is not an official website to run Autoprefixer. It has many small issues :D.

**Comment by franktopel** [comment]
This is leading away from the original problem.

**Comment by ai** [comment]
> This is leading away from the original problem.

Try `postcss-cli` it will show you a warning (sorry, I am on vacation, can’t say what you need to fix in grid)

**Comment by franktopel** [comment]
I know the css grid has been an ongoing topic for a long time, but in our company we're going to have a grand meeting tomorrow in which we are to decide whether to use standard based CSS and rely on autoprefixer or whether to continue generating css rules in Javascript for IE 11. I'd love to say "we can rely on autoprefixer".

**Comment by franktopel** [comment]
This is what postcss-cli generates from `a.css`
```
.a { grid-row-span: 4; }
.b { grid-col-span: 4; }
```
```
@connexo > postcss a.css -o b.css
.a { grid-row-span: 4; }
.b { grid-col-span: 4; }
```

**Comment by ai** [comment]
If you want to run Autoprefixer on your existed CSS — it i bad idea. You must create grids from the beginning with Autoprefixer limits in the minds (this is why this feature is under the flag).

Also, grid by classes (like `colspan1`) is a bad idea. Grid layout has very useful `grid-template` feature. And we created Grid feature in Autoprefixer by looking to this way of using it. I really recommend describing styles in CSS, not in HTML by classes.

So, Autoprefixer grids are not “out of box” feature. You will need to learn how to use it (article about it will be posted in the middle of July by @Dan503).

**Comment by ai** [comment]
@franktopel show me your `postcss.config.js` which you used in `postcss-cli`

**Comment by franktopel** [comment]
I did not configure any, I just installed postcss-cli and autoprefixer globally. Where do I put it and what do I put in it for testing?

**Comment by ai** [comment]
@franktopel you must set `grid: true` option. So you must create `postcss.config.js` (in the same dir, where you put `a.css`):

```js
module.exports = {
  plugins: [
    require('autoprefixer')({ grid: true })
  ]
}
```

**Comment by franktopel** [comment]
*I really recommend describing styles in CSS, not in HTML by classes.* I agree under normal circumstances, but I cannot agree in this case with the desribed "layout must be created using generation process" case.

**Comment by franktopel** [comment]
Using your config the result stays the same:

```
.a { grid-row-span: 4; }
.b { grid-col-span: 4; }
```

**Comment by ai** [comment]
If you want to create `.colspan1`, you don’t need an Autoprefixer. Just put `-ms-` properties manually to mixin. It is a single place, so you can describe all hacks there and then do not think about it.

**Comment by franktopel** [comment]
That might be a workaround, but question remains why are you not pre-fixing these basic properties? Is `grid-col-span` not properly prefixed when you generate `-ms-grid-col-span`?

**Comment by ai** [comment]
Wait a second. What is `grid-row-span`? I can’t find it in [CSS grid spec](https://www.w3.org/TR/css-grid-1/)

**Comment by franktopel** [comment]
https://css-tricks.com/almanac/properties/g/grid-row-column-span/

**Comment by Dan503** [comment]
Go with the IE version and auto prefixer.

I've got a three part series coming out in just 2 or 3 weeks that will teach you everything you need to know about using css-grid safely in IE.

The issue you are facing is that you are not specifying a start column or an end column. You can't just specify a span and leave it at that for Autoprefixer to work.

Here is the bit in the article about spanning multiple cells:

------

Autoprefixer has limited column and row spanning support

There is only one way to span multiple cells in IE. You need to define two things, a starting cell, and how many other cells you would like to span.


    .grid-cell {
      -ms-grid-column: 1; /* starting cell */
      -ms-grid-column-span: 2; /* number of cells to span */
    }

In modern browsers, you have access to far more options.

Autoprefixer friendly

Out of the modern ways to span multiple cells, Autoprefixer fully supports the following. Feel free to use any of these methods as much as you like:

Specify a starting line and the number of lines to span (similar to IE):


    .grid-cell {
      grid-column: 1 / span 2;
    }

Specify an end line and then span backwards:


    .grid-cell {
      grid-column: span 2 / 3;
    }

Specify a starting line and an end line directly:


    .grid-cell {
      grid-column: 1 / 3;
    }


Autoprefixer unfriendly. Here be dragons!

Now this is where Autoprefixer reaches its limit. The following methods are supported in modern browsers but are not supported by Autoprefixer. This is mainly due to Autoprefixer having no idea what grid the grid cell belongs to. For the last one, it is really only useful if you have access to auto-placement which is not something that IE supports.

Specify a starting line and how many lines from the end of the explicit grid to span:


    .grid-cell {
      grid-column: 1 / -1;
    }
    /* also works in reverse */
    .grid-cell {
      grid-column: -1 / 1;
    }

Specify both the start and end line from the end of the explicit grid:


    .grid-cell {
      grid-column: -3 / -1;
    }

Specify only how many lines to span:


    .grid-cell {
      grid-column: span 2;
    }

So basically avoid using negative integers and don’t use purely just a span and you will be fine :)

**Comment by franktopel** [comment]
In the page I linked in my previous comment someone commented: https://css-tricks.com/almanac/properties/g/grid-row-column-span/#comment-1610564


*Justin McDowell

For the record, you can now accomplish this using the grid-column-end property instead:

grid-column-end: span 3;

See it in action here: https://codepen.io/ragdoll/pen/mMbrNj*

**Comment by ai** [comment]
This article is old and it says wrong things. The final Grid spec uses different syntax.

I will add warning for this case since this bad article can be popular.

Will you create a issue in Vue?

**Comment by ai** [comment]
Yeap `grid-column-end` should work (but I don't have my laptop right now to test).

**Comment by franktopel** [comment]
This is the JS snippet that has been used so far to create the grid classes:

```
var style = document.getElementById("css-grid").innerHTML;
var cols = 12;
var rows = 50;
var col;
for (col = 1; col < cols + 1; col++) {
                style = style + "\n.grid" + col + " { display: -ms-grid; -ms-grid-columns: (1fr)[" + col + "]; -ms-grid-rows: auto; }";
                style = style + "\n.col" + col + " { -ms-grid-column: " + col + "; }";
}
var row;
for (row = 1; row < rows + 1; row++) {
                style = style + "\n.row" + row + " { -ms-grid-row: " + row + "; }";
}
var colspan;
for (colspan = 1; colspan < cols + 1; colspan++) {
                style = style + "\n.colspan" + colspan + " { -ms-grid-column-span: "
                                               + colspan + "; }";
}
var rowspan;
for (rowspan = 1; rowspan < rows + 1; rowspan++) {
                style = style + "\n.rowspan" + rowspan + " { -ms-grid-row-span: "
                                               + rowspan + "; }";
}
```

And these are examples of how they generated their HTML:

`<div class="col1 row1 colspan1 rowspan1">
  <div class="bm-header">
    Code-Table
  </div>
</div>`

**Comment by franktopel** [comment]
>Will you create a issue in Vue?

For what? The postcss-cli doesn't issue a warning for this test case either.

**Comment by YozhikM** [comment]
@franktopel Here is a [discussion](https://github.com/csstools/postcss-preset-env/pull/31#issuecomment-391000277). Including why you can not trust the authors of articles, especially older articles

**Comment by ai** [comment]
I think it is much better to use class like `header` in HTML and describe position of element in CSS by `grid-template-areas`. Describing position by classes is bad since it doesn't work good with media queries.

**Comment by franktopel** [comment]
@ai Again, I agree, but their layouts need to be generated from layout.xml files used by the Swing client. Manually translating those into working HTML layouts would take like 50000 man days.

**Comment by ai** [comment]
> For what? The postcss-cli doesn't issue a warning for this test case either.

You tested Vue with another CSS which must show warning.

You tested `postcss-cli` only with wrong CSS. This is why it didn't show you a warning.

**Comment by franktopel** [comment]
@ai Will post that over there with this test:

```
.warn {
    grid-column-end: 3;
    grid: subgrid;
        grid-template-areas: "head head"
                         "nav  main"
                         "foot ....";
}
```

**Comment by ai** [comment]
Yeap. And mention me there 😊. I will tell them what to fix.

**Comment by franktopel** [comment]
Maybe this would have been the more appropriate place to raise the issue: https://github.com/postcss/postcss-loader

**Comment by franktopel** [comment]
Aside from that issue, what are you going to do about autoprefixer & grid-col-span / grid-row-span?

**Comment by ai** [comment]
> Aside from that issue, what are you going to do about autoprefixer & grid-col-span / grid-row-span

Try to find more recent article about Grid with final syntax. I think you need to use `grid-row-start`/`grid-row-end` (but without a laptop I can't say you exact code to fix your mixins).

**Comment by franktopel** [comment]
> Go with the IE version and auto prefixer. 

@Dan503 Please look at the Javascript snippet and the HTML example I posted above. Is there even a cross-browser solution possible if the `.rowspan1` class never mentions where the cell starts?

**Comment by Dan503** [comment]
> Is there even a cross-browser solution possible if the `.rowspan1` class never mentions where the cell starts?

If the statement doesn't mention either a start or an end point then no, Autoprefixer can't save you.

**Comment by franktopel** [comment]
So the approach would be:

`<div class="col1 row1 colspan4 rowspan20"></div>`

would need additional CSS-classes

<div class="col4 row7 colspan4 rowspan20 colend9 rowend24"></div>`

with 

```
.colend9 {
  grid-column-end: span 9 /* colx + colspany */;
}
```

and
```
.rowend24 {
  grid-row-end: span 24 /* rowx + rowspany */;
}
```

or how would you advise approaching this?

**Comment by franktopel** [comment]
These are the other classes generated by that `grid-create.js`:

```
.grid1 { display: -ms-grid; -ms-grid-columns: (1fr)[1]; -ms-grid-rows: auto; }
.col1 { -ms-grid-column: 1; }
.grid2 { display: -ms-grid; -ms-grid-columns: (1fr)[2]; -ms-grid-rows: auto; }
.col2 { -ms-grid-column: 2; }
.grid3 { display: -ms-grid; -ms-grid-columns: (1fr)[3]; -ms-grid-rows: auto; }
.col3 { -ms-grid-column: 3; }
.grid4 { display: -ms-grid; -ms-grid-columns: (1fr)[4]; -ms-grid-rows: auto; }
.col4 { -ms-grid-column: 4; }
.grid5 { display: -ms-grid; -ms-grid-columns: (1fr)[5]; -ms-grid-rows: auto; }
.col5 { -ms-grid-column: 5; }
.grid6 { display: -ms-grid; -ms-grid-columns: (1fr)[6]; -ms-grid-rows: auto; }
.col6 { -ms-grid-column: 6; }
.grid7 { display: -ms-grid; -ms-grid-columns: (1fr)[7]; -ms-grid-rows: auto; }
.col7 { -ms-grid-column: 7; }
.grid8 { display: -ms-grid; -ms-grid-columns: (1fr)[8]; -ms-grid-rows: auto; }
.col8 { -ms-grid-column: 8; }
.grid9 { display: -ms-grid; -ms-grid-columns: (1fr)[9]; -ms-grid-rows: auto; }
.col9 { -ms-grid-column: 9; }
.grid10 { display: -ms-grid; -ms-grid-columns: (1fr)[10]; -ms-grid-rows: auto; }
.col10 { -ms-grid-column: 10; }
.grid11 { display: -ms-grid; -ms-grid-columns: (1fr)[11]; -ms-grid-rows: auto; }
.col11 { -ms-grid-column: 11; }
.grid12 { display: -ms-grid; -ms-grid-columns: (1fr)[12]; -ms-grid-rows: auto; }
.col12 { -ms-grid-column: 12; }
.row1 { -ms-grid-row: 1; }
.row2 { -ms-grid-row: 2; }
.row3 { -ms-grid-row: 3; }
.row4 { -ms-grid-row: 4; }
.row5 { -ms-grid-row: 5; }
.row6 { -ms-grid-row: 6; }
.row7 { -ms-grid-row: 7; }
.row8 { -ms-grid-row: 8; }
.row9 { -ms-grid-row: 9; }
.row10 { -ms-grid-row: 10; }
/* ... 11 - 47 */
.row48 { -ms-grid-row: 48; }
.row49 { -ms-grid-row: 49; }
.row50 { -ms-grid-row: 50; }
```

**Comment by Dan503** [comment]
That approach can work if you are able to generate the css as you work with the html rather than writing the css before hand.

In order to make that system work with Autoprefixer you would need to write styles like this.

````css
.col4.colspan5 {
  grid-column: 4 / span 5;
}
row7.rowspan20  {
  grid-row: 7 / span 20;
}
````

It would be impossible to pre-write the css because there would be infinite possible combinations.

**Comment by Dan503** [comment]
A cell can either span a number of columns or end at a certain line number.

If you say that a line starts at line 1, spans only 2 cells but some how ends at line 10... how exactly is that logically supposed to work? Does it end at line 10 or only span 2 cells? It's impossible to do both at the same time.

**Comment by franktopel** [comment]
I don't get where you see that problem in the mentioned approach?

**Comment by Dan503** [comment]
Actually I think this would work except you would need to do all of the IE stuff manually. Autoprefixer won't help you. 

````
:root {
  --col-span: 1;
  --row-span: 1;
}
.col4 {
  -ms-grid-column: 4;
  grid-column: 4 / span var(--span);
}
.row4 {
  -ms-grid-column: 4;
  grid-column: 4 / span var(--span);
}
.colspan5 {
 -ms-grid-column-span: 5;
 --col-span: 5;
}
.rowspan20 {
 -ms-grid-row-span: 20;
 --row-span: 20;
}
````
That approach uses CSS variables for the modern browsers and IE syntax for IE.

**Comment by franktopel** [comment]
My approach is to let the generator tool calculate the appropriate end positions and use static CSS classes instead of using CSS variables.

**Comment by franktopel** [comment]
@ai This still needs your attention and some form of handling, even if that's only a warning that tells the user `grid-column-span` and `grid-row-span` have been removed from the specs.

**Comment by ai** [comment]
Oh, sure

**Comment by franktopel** [comment]
@Dan503 Does that mean that what is now `grid-column-start` would be `-ms-grid-column` for IE11?

**Comment by Dan503** [comment]
Yes

**Comment by franktopel** [comment]
Well, then I get it. I like your approach alot!

**Comment by franktopel** [comment]
So based on that, would this be the correct adjustment of the mixins I wrote to replace `grid-create.js`:

```
$columns: 12;
$rows: 50;

:root {
  --col-span: 1;
  --row-span: 1;
}

@mixin col-x {
  @for $i from 1 through $columns {
  .col#{$i} { -ms-grid-column: $i; grid-column: $i / span var(---col-span); }
  }
}

@mixin grid-x {
  @for $i from 1 through $columns {
  .grid#{$i} { display: grid; grid-template-columns: (1fr)[$i]; }
  }
}

@mixin row-x {
  @for $i from 1 through $rows {
  .row#{$i} { -ms-grid-row: $i; grid-row: $i / span var(---row-span); }
  }
}

@mixin colspan-x {
  @for $i from 1 through $columns {
  .colspan#{$i} { -ms-grid-column-span: $i; --col-span: $i; }
  }
}

@mixin rowspan-x {
  @for $i from 1 through $rows {
  .rowspan#{$i} { -ms-grid-row-span: $i; --row-span: $i; }
  }
}

@include col-x;
@include grid-x;
@include row-x;
@include colspan-x;
@include rowspan-x;
```

**Comment by franktopel** [comment]
Unfortunately, this is what **dartsass** generates from this:

```
.colspan9 {
  -ms-grid-column-span: 9;
  --col-span: $i;
}
```

Any idea?

**Comment by franktopel** [comment]
Solved: For some reason I don't get `$i` has to be written as `#{$i}` when used in a css variable:

```
@mixin colspan-x {
  @for $i from 1 through $columns {
  .colspan#{$i} { -ms-grid-column-span: $i; --col-span: #{$i}; }
  }
}

@mixin rowspan-x {
  @for $i from 1 through $rows {
  .rowspan#{$i} { -ms-grid-row-span: $i; --row-span: #{$i}; }
  }
}
```

**Comment by Dan503** [comment]
Yeah that looks pretty good I think except for this bit:
```scss
@mixin grid-x {
  @for $i from 1 through $columns {
  .grid#{$i} { display: grid; grid-template-columns: (1fr)[$i]; }
  }
}
```
It should be this:
```scss
@mixin grid {
    .grid {
      display: -ms-grid;
      display: grid;
      -ms-grid-columns: (1fr)[$columns]; 
      grid-template-columns: repeat($columns, 1fr); 
    }
}
```

**Comment by franktopel** [comment]
Ah, I was wondering for some time what the number in square brackets means. Thanks alot!

**Comment by Dan503** [comment]
Actually instead of `1fr` use `minmax(0, 1fr)` it will work more like how you expect it to work.

```
 -ms-grid-columns: (minmax(0,1fr))[$columns]; 
      grid-template-columns: repeat($columns, minmax(0,1fr)); 
```

**Comment by franktopel** [comment]
Please elaborate?

**Comment by franktopel** [comment]
`display: grid` will still be handled by autoprefixer, no?

**Comment by franktopel** [comment]
Ah, I see. It's probably a good idea not to have autoprefixer handle this part of our CSS at all. I supposed we better turn off grid in the .postcssrc.js config file.

**Comment by franktopel** [comment]
@Dan503 You probably meant:

```
@mixin grid-x {
    .grid#{$i} {
      display: -ms-grid;
      display: grid;
      -ms-grid-columns: (minmax(0,1fr))[$columns]; 
      grid-template-columns: repeat($columns, minmax(0,1fr)); 
    }
}
```
right?

**Comment by Dan503** [comment]
Yeah `display: grid ` is translated by Autoprefixer. I wasn't sure if you were still going to turn the grid setting on or not.

`1fr` = `minmax(min-content, 1fr)`

What this means is that if you have something like a wide table in your grid, it will actually expand the grid to fit the table. The columns will not be all equal widths like you would expect them to be. The columns holding the table will be wider than the columns not holding the table. If the screen is small enough, the grid will overflow it's container (possibly going off the screen) rather than shrink the grid cells at a consistent rate to fit inside it's container.

**Comment by franktopel** [comment]
We do have tables with 70 or 80 columns, those are meant to be horizontally scrolled.

**Comment by Dan503** [comment]
Yes you will most definitly want to use `minmax(0,1fr)`. Using `1fr` would mean that the table overflow functionality would never come into effect.

**Comment by Dan503** [comment]
> So in those cases your suggestion would destroy the layout?

Only if you use `1fr`. If you use `minmax(0,1fr)` it won't cause the issue.

**Comment by franktopel** [comment]
Is `minmax` a browser function, or scss function?

**Comment by ai** [comment]
@franktopel [browser](https://developer.mozilla.org/ru/docs/Web/CSS/minmax)

**Comment by franktopel** [comment]
Thanks. Thankfully it is supported by IE 11 as well: https://caniuse.com/#search=minmax

**Comment by Dan503** [comment]
Yeah I wouldn't have recommended using `minmax()` if it wasn't natively supported in IE.

**Comment by Dan503** [comment]
You actually don't need to place any of that code into mixins. You aren't going to be calling them again anywhere.

Or are you using the mixins just as a way to help organise the code into easily recognisable chunks?

**Comment by franktopel** [comment]
Yeah, I find it easier to clearly structure and name what is going on.

**Comment by franktopel** [comment]
Just to make sure, you haven't removed the loop from grid-x mixin intentionally, have you?

**Comment by Dan503** [comment]
@franktopel 
> You probably meant: ... (code with #{$i} in it) ...

No, I meant what I wrote. You don't need the `#{$i}` bit. I intentionally removed it.

You aren't looping over anything. You only need to state the .grid rule once. There is no `$i` at that point in time in the code.

**Comment by Dan503** [comment]
> Just to make sure, you haven't removed the loop from grid-x mixin intentionally, have you? 

Yes I intentionally removed the loop. You don't need it and shouldn't use it.

**Comment by franktopel** [comment]
I see, but our generator generates classes `.grid1` through `.grid12`. I don't want to use a selector like `[class*=grid]` in our css, and I'm really glad no generator changes will be necessary.

**Comment by Dan503** [comment]
Look at the code you wrote. The only thing that changes between each iteration is what the grid class is.

The loop would result in 12 identical .grid rules whose only difference is a number at the end of the class name.

**Comment by franktopel** [comment]
This was the old mixin:

```scss
@mixin grid-x {
  @for $i from 1 through $columns {
  .grid#{$i} { display: grid; grid-template-columns: (1fr)[$i]; }
  }
}
```

**Comment by Dan503** [comment]
Ok I've got a better idea, rather than generate the whole rule 12 times, use `@extends`

```scss
    %grid {
      display: -ms-grid;
      display: grid;
      -ms-grid-columns: (1fr)[$columns]; 
      grid-template-columns: repeat($columns, 1fr); 
    }

  @for $i from 1 through $columns {
  .grid#{$i} { @extends %grid; }
  }
```

**Comment by franktopel** [comment]
That doesn't look like working syntax to me. You're not passing `$i` to `%grid`.

**Comment by franktopel** [comment]
```scss
@mixin grid-x {
  @for $i from 1 through $columns {
    .grid#{$i} {
      display: -ms-grid;
      display: grid;
      -ms-grid-columns: (minmax(0, 1fr)) [$i];
      grid-template-columns: repeat($i, minmax(0, 1fr));
    }
  }
}
```

I'm getting dizzy. Is the above it?

**Comment by franktopel** [comment]
```scss
$columns: 12;
@mixin grid-x {
  %grid {
    display: -ms-grid;
    display: grid;
    -ms-grid-columns: (minmax(0,1fr))[$columns]; 
    grid-template-columns: repeat($columns, minmax(0,1fr)); 
  }

  @for $i from 1 through $columns {
    .grid#{$i} { @extends %grid; }
  }
}
@include grid-x;```

doesn't work (check sassmeister.com).

It generates to 

```css
.grid1 {
  @extends %grid;
}

.grid2 {
  @extends %grid;
}

.grid3 {
  @extends %grid;
}

.grid4 {
  @extends %grid;
}

.grid5 {
  @extends %grid;
}

.grid6 {
  @extends %grid;
}

.grid7 {
  @extends %grid;
}

.grid8 {
  @extends %grid;
}

.grid9 {
  @extends %grid;
}

.grid10 {
  @extends %grid;
}

.grid11 {
  @extends %grid;
}

.grid12 {
  @extends %grid;
}
```

**Comment by Dan503** [comment]
... maybe it's `@extend`

**Comment by franktopel** [comment]
If you correct `@extends` to `@extend` this is the result:

```css
.grid12, .grid11, .grid10, .grid9, .grid8, .grid7, .grid6, .grid5, .grid4, .grid3, .grid2, .grid1 {
  display: -ms-grid;
  display: grid;
  -ms-grid-columns: minmax(0, 1fr) [12];
  grid-template-columns: repeat(12, minmax(0, 1fr));
}
```

This is not what the original code created - this is twelve rules with identical content. $columns inside the loop must be $i.

**Comment by Dan503** [comment]
Oh thought you wanted a 12 column grid.

I didn't realise you wanted 2 column, 3 column, 4 column (etc.) grid options :/

**Comment by franktopel** [comment]
I think that's being used for sub-grids.

**Comment by franktopel** [comment]
So is this it?

```scss
@mixin grid-x {
  @for $i from 1 through $columns {
    .grid#{$i} {
      display: -ms-grid;
      display: grid;
      -ms-grid-columns: (minmax(0, 1fr)) [$i];
      grid-template-columns: repeat($i, minmax(0, 1fr));
    }
  }
}
```

**Comment by Dan503** [comment]
Yes, that is the code you want to use (Sorry about the confusion)

Be aware the the grid cells will not have gaps though. They will all be hard up against one another.

Adding gaps will be a lot harder since IE doesn't support them :(

**Comment by franktopel** [comment]
Gaps are waste of space anyway, right? :D

**Comment by Dan503** [comment]
It would still be possible. It means having to take extra columns into account when calculating the math.

**Comment by Dan503** [comment]
@ai (Sorry about all the spam)

**Comment by franktopel** [comment]
This would be the version using `@include` which would allow for reuse:

```scss
$columns: 12;
@mixin grid-x($cols) {
  .grid#{$cols} {
    display: -ms-grid;
    display: grid;
    -ms-grid-columns: (minmax(0, 1fr))[$cols];
    grid-template-columns: repeat($cols, minmax(0, 1fr));
  }
}
@mixin grid {
  @for $i from 1 through $columns {
    @include grid-x($i);  
  }
}
@include grid;
```

**Comment by Dan503** [comment]
I think it can be even nicer

```scss
@mixin grid-x($cols) {
  .grid#{$cols} {
    display: -ms-grid;
    display: grid;
    -ms-grid-columns: (minmax(0, 1fr))[$cols];
    grid-template-columns: repeat($cols, minmax(0, 1fr));
  }
}
@mixin grid($columns) {
  @for $i from 1 through $columns {
    @include grid-x($i);  
  }
}
@include grid(12);
```

**Comment by Dan503** [comment]
That way it defaults to 12 but you can use a different number if you want.

**Comment by franktopel** [comment]
I'm using `$columns` in two more mixins - so I don't see much of an improvement there.

**Comment by Dan503** [comment]
Ok yeah, fair point.

**Comment by Dan503** [comment]
> Tbh, I don't see an advantage in splitting this up into two mixins.

Yeah there isn't much extra advantage. They can just all be in the one mixin.

**Comment by ai** [comment]
Warning was added f8ce7b3

**Comment by franktopel** [comment]
@ai The deeper I dive into it, the more I ask myself why that shouldn't be prefixable.

```css
grid-column: 1 / span 6;
```
respectively
```
grid-column-start: 1;
grid-column-end: span 6;
```
can be prefixed to
```css
-ms-grid-column: 1;
-ms-grid-column-span: 6;
```
Why would that not work?

**Comment by ai** [comment]
@franktopel `autoprefixer({ grid: true })` converts

```css
.foo {
  grid-column: 1 / span 6;
}
```

to

```css
.foo {
  -ms-grid-column: 1;
  -ms-grid-column-span: 6;
  grid-column: 1 / span 6;
}
```

Why do you think it doesn’t work?

**Comment by franktopel** [comment]
Just ignore my comment, I deleted it for a reason. I've just confused the properties for the ones that were the reason for this issue.

**Comment by Dan503** [comment]
Actually @franktopel has a point.

````
.foo {
  grid-column: 1 / span 6;
}
````
is shorthand for this
````
.foo {
  grid-column-start: 1;
  grid-column-end: span 6;
}
````
So it could translate that into this:
````
.foo {
  -ms-grid-column: 1;
  grid-column-start: 1;
  -ms-grid-column-span: 6;
  grid-column-end: span 6;
}
````
It could also turn this:
````
.foo {
  grid-column-start: 1;
}
.bar {
  grid-column-end: span 6;
}
````
Into this quite safely
````
.foo {
  -ms-grid-column: 1;
  grid-column-start: 1;
}
.bar {
  -ms-grid-column-span: 6;
  grid-column-end: span 6;
}
````
Where Autoprefixer would reach it's limits though is in these examples (these aren't able to be translated)
```
.foo {
  // IE can't span backwards and Autoprefixer doesn't know the end value
  grid-column-start: span 6;
}
.bar {
  //IE can't do "end" and Autoprefixer doesn't know the start value
  grid-column-end: 3;
}
```

**Comment by ai** [comment]
@Dan503 I don’t get you ;).

What is CSS input, current output, and expected output?

**Comment by Dan503** [comment]
Nevermind the Autoprefixer safe stuff is outputting exactly as expected.

The Autoprefixer unsafe stuff isn't outputting/erroring as expected though.

Input:
````css
.xxx {
  /* IE can't span backwards and Autoprefixer doesn't know the end value */
  grid-column-start: span 6;
}

.yyy {
  /* IE can't do "end" and Autoprefixer doesn't know the start value */
  grid-column-end: 3;
}
````

Output
````css
.xxx {
  /* -ms-grid-column: span 6; is invalid css in IE */
  -ms-grid-column: span 6;
      grid-column-start: span 6;
}
.yyy {
  /* -ms-grid-column-span: 3; != grid-column-end: 3; */
  /* IE can't do "end" and Autoprefixer doesn't know the start value */
  -ms-grid-column-span: 3;
  grid-column-end: 3;
}
````

**Comment by ai** [comment]
@Dan503 can you open a new issue? We will prevent `-ms-` prefix and will show warning.

**Comment by Dan503** [comment]
New issue created here: https://github.com/postcss/autoprefixer/issues/1064

**Comment by franktopel** [comment]
@Dan503 We might want to find a channel to communicate other than this issue tracker ;)

Why does IE in this example render all `.box` grid-items above one another? Which -ms property am I missing here for IE to create a flow of grid-items?

https://codepen.io/connexo/pen/QxOeVx?editors=1100

**Comment by Dan503** [comment]
@franktopel You haven't added column classes to any of the boxes. IE can't do autoplacement. every grid cell in IE must be placed manually on the grid with column and row coordinates

**Comment by franktopel** [comment]
You got to be joking...

**Comment by Dan503** [comment]
No seriously, in IE there is no Autoplacement of any kind whatsoever. Every cell needs to be placed manually.

That is why the Autoprefixer grid areas are so useful. You can give each cell an area to occupy then let autoprefixer recalculate all of the cell positions when using media queries.

**Comment by Dan503** [comment]
@franktopel
> We might want to find a channel to communicate other than this issue tracker ;) Any suggestions?

You can [follow and DM me on twitter](https://twitter.com/Daniel_Tonon) ;)

**Comment by franktopel** [comment]
@Dan503 Unfortunately I'm not using twitter or any other social network :(

**Comment by Dan503** [comment]
@franktopel my guide to using CSS grid in IE has come out 😁

Part 1
https://css-tricks.com/css-grid-in-ie-debunking-common-ie-grid-misconceptions/

Part 2
https://css-tricks.com/css-grid-in-ie-css-grid-and-the-new-autoprefixer/

Part 3 (coming soon)
